### ISSUE TITLE
Issue rendering images via <Image> component

### ISSUE THREAD

**Issue by grantburnham** [diegomura/react-pdf#929]
I have been unable to correctly load images via the <Image src /> component in React PDF.

I had issue locally in my code but have also been able to replicate the same issue on the image test page
https://react-pdf.org/repl?example=images

This works and renders (test image)
`https://upload.wikimedia.org/wikipedia/en/a/a9/Example.jpg`
<Image src="https://upload.wikimedia.org/wikipedia/en/a/a9/Example.jpg" />

This will not render (test image)
`https://ssl.gstatic.com/ui/v1/icons/mail/rfr/logo_gmail_lockup_default_1x.png`
<Image src="https://ssl.gstatic.com/ui/v1/icons/mail/rfr/logo_gmail_lockup_default_1x.png" />

Is there something that the <Image> component needs to support this?

The images I wish to load for my project will be PNG but these look to be supported.

Any help would be greatly appreciated

**Comment by kaloraat** [comment]
I am having same issue. Can't display image at all. 
Tried the docs but its very confusing. Where how are we supposed to use this in Image component?
From the [docs](https://react-pdf.org/components#source-object):
`{ uri: valid-url, method: 'GET', headers: {}, body: '' }`
![Screen Shot 2020-06-20 at 12 00 16 AM](https://user-images.githubusercontent.com/15702617/85140794-1e7d1900-b289-11ea-8b10-b8172aac9947.png)
This is Image componet. What does docs trying to say??
<Image  />
Use this way. Of course it does not work.
`<Image { uri: valid-url, method: 'GET', headers: {}, body: '' }` />`

**Comment by studio-blueboat** [comment]
I'm also having this issue. 
I can load the wikipedia image you've mentioned above @grantburnham through the src prop but am unable to load another external image, say this one. https://positivewomen.org.au/wp-content/uploads/2019/11/PWV_hero4.5_sm.jpg

**Comment by grantburnham** [comment]
@studio-blueboat I know, seems illogical really, as I can't see anything that should stop it from loading and nothing in the docs seems to point to any restrictions.

**Comment by jyenes** [comment]
I have exactly the same issue trying to load images from firebase, only local ones work.

**Comment by vsachini** [comment]
I'm having the same issue here.

**Comment by jyenes** [comment]
I wonder if there is any way to pre-load the images on memory to workaround this issue.

**Comment by sterp** [comment]
May be just me but I'm having issues using both PDFViewer or download with local and remote images. Haven't tried Firebase hosted images.

**Comment by jyenes** [comment]
I have downgraded the version to 1.6.9 and I managed to get the images on the PDF, hope that help

**Comment by studio-blueboat** [comment]
@grantburnham I'm also having issues with local .png images, specifically the following image and others rendered out of AI in a similar manner.

![speech-bubble](https://user-images.githubusercontent.com/50648946/85805168-eb59ee80-b78e-11ea-88ef-2889ef7d7206.png)

**Comment by danielbyun** [comment]
adding `?noCache={randomString}` and configuring CORS does not work for me for rendering S3 images. Trying to figure out if there's anything the server can do regarding the signature and the key when sending the S3 URL.

**Comment by grantburnham** [comment]
@studio-blueboat I also thought it may be to do with image type, but docs look to support a good image range. I tried local and remote .jpg, .png, but can rarely get one to display, again tried loading local image but this still wouldn't render

**Comment by grantburnham** [comment]
I have noticed that images I am trying to load which are failing seem to be failing due to CORS (see loading google image below via console logs)

`Access to XMLHttpRequest at 'https://ssl.gstatic.com/ui/v1/icons/mail/rfr/logo_gmail_lockup_default_1x.png' from origin 'https://react-pdf.org' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
`
The Wikipedia images isn't throwing the same error and will load successfully.

This may be the issue why we can't see the images, maybe we can see others are finding the same.

**Comment by Jakku7** [comment]
Anyone was able to fix it or find a workaround? I tried downgrading to version 1.6.9 but it still doesn't work.

**Comment by roshangm1** [comment]
This has been here for a while and it's the issue with react-pdf. @diegomura

**Comment by Jakku7** [comment]
> I am having same issue. Can't display image at all.
> Tried the docs but its very confusing. Where how are we supposed to use this in Image component?
> From the docs:
> { uri: valid-url, method: 'GET', headers: {}, body: '' }
> 
> This is Image componet. What does docs trying to say??
> 
> Use this way. Of course it does not work.
> <Image { uri: valid-url, method: 'GET', headers: {}, body: '' } />`

You need to pass an object to the src/source prop. So in this case it should be like that: 
`<Image src={ uri: "", method: "", headers: "", body: "" } />`

It's better to pass a function that would return an object though. Like that: 
```
<Image src={ returnSrc()} />

function returnSrc() {
  const path = "https://upload.wikimedia.org/wikipedia/en/a/a9/Example.jpg" as const;
  const selectedMethod = 'GET' as const;
  return { uri: path, method: selectedMethod, body: '', headers: '' }; 
}
```

I figured out that the bug lies in the image size. Source is not important. I uploaded Wikipedia image on S3 and it worked. Than I resized one of the images that didn't work previously to 800x600 and it also started to work.

**Comment by danielbyun** [comment]
@Jakku7 

That's very interesting... I copied and hard-coded one of the pre-signed URLs from the S3 bucket and was able to render it no matter what the size was... but as soon as I make it so that it's a URL it seems to break.

**Comment by roshangm1** [comment]
Also, I was trying this with firebase hosted images. It wasn't working at all. And then installed this extension called "CORS Everywhere" in firebase. After enabling that, it started working. But configuring same cors setting in firebase.json doesn't work.

**Comment by roshangm1** [comment]
Another finding, along with the size of the image, I can confirm that it's issue with the CORS. So, firebase.json didn't work but for images in firebase storage, this worked. Configuring cors for gcloud storage. 

https://firebase.google.com/docs/storage/web/download-files#cors_configuration

**Comment by jamigibbs** [comment]
I'm running into this same issue with an S3 hosted image. It's definitely a CORS issue. I'm not able to configure the server to allow for this to work so the only other way I'm seeing is to proxy the request with my own server. You can test this out with something like [cors anywhere](https://github.com/Rob--W/cors-anywhere/) and route your image like this:

https://cors-anywhere.herokuapp.com/http://example.com/your-image.png

**Comment by AwolDes** [comment]
@jamigibbs did [S3s CORS policies](https://docs.aws.amazon.com/AmazonS3/latest/dev/cors.html) help at all here? Having a similar issue to what you're experiencing and wondering if it's worth a shot

**Comment by jamigibbs** [comment]
@AwolDes In my case I did not have image files to upload. I needed to use a direct link to an S3 server where they lived and I did not have access to those settings either. Sounds like a good workaround in your case though.

**Comment by danielbyun** [comment]
@Jakku7 's solution above worked for me! Made the bucket public for CORs and set the uri to the full image's URL.

**Comment by alanmaranto** [comment]
> @jamigibbs did [S3s CORS policies](https://docs.aws.amazon.com/AmazonS3/latest/dev/cors.html) help at all here? Having a similar issue to what you're experiencing and wondering if it's worth a shot
> 
> Edit: Success! 🎉
> 
> I created a bucket with public access (it will act as a CDN), uploaded my image and making sure it was publicly accessible.
> 
> Then I added the following CORS configuration
> ![image](https://user-images.githubusercontent.com/9693812/91154445-e2d95200-e704-11ea-9d90-0dc5e39d1d32.png)
> 
> ```
> <CORSConfiguration>
>  <CORSRule>
>    <AllowedOrigin>*</AllowedOrigin>
>    <AllowedMethod>GET</AllowedMethod>
>  </CORSRule>
> </CORSConfiguration>
> ```
> 
> Worked like a charm.

This worked perfect for me !!
Thanks

**Comment by SH-Stark** [comment]
Any workaround without granting public access ? I need to display a signature of a user... that should be accessible only for the user.
Thanks

**Comment by diegomura** [comment]
Seems to me that most of the issues here are CORS related right? There's no much to do from my side rather than allowing to pass custom headers as this lib currently does. Correct me if I'm wrong though! Anyone with CORS issues should make sure that the client can actually access those resources. I'm closing this in some days if there's no other thing going on or if happens that there's nothing to do from this lib side. Thanks!

**Comment by reinvanimschoot** [comment]
Can someone explain to why a normal `<img>` works fine but an `<Image>` component gives us errors? I really want to understand where the trouble lies exactly. Thanks!

**Comment by rvanpraet** [comment]
> 
> 
> Seems to me that most of the issues here are CORS related right? There's no much to do from my side rather than allowing to pass custom headers as this lib currently does. Correct me if I'm wrong though! Anyone with CORS issues should make sure that the client can actually access those resources. I'm closing this in some days if there's no other thing going on or if happens that there's nothing to do from this lib side. Thanks!

We have an issue where we have set CORS fully open in our Azure Blob Storage, making it possible to directly access the images from the web browser without any custom headers. 

![afbeelding](https://user-images.githubusercontent.com/43333514/118466988-aec71680-b703-11eb-8eb9-35124690ab33.png)

Passing the url via the library however, will give the CORS error described in this report. So it does seem related to the library, although I'm also in the dark to why this could happen..

I've also tested with public images from https://picsum.photos/ , these seem to work until the size exceeds 800x600 as mentioned in the comment of @Jakku7 

Hope you can further look into this issue !

**Comment by himanshusandha** [comment]
I had the same CORS issue while accessing image stored in S3 bucket, I was able to access the image in normal Image tag of browser but when the React-PDF tried to render it was throwing CORS error. Passing `"Cache-Control": "no-cache"` in Header worked for me

i.e while mentioning URL in Image tab of React-PDF instead of assigning URL directly to `src` attribute I made call as below
`<Image src={{ uri: _your_image_url_goes_here_, method: "GET", headers: { "Cache-Control": "no-cache" }, body: "" }} />` and it worked.

**Comment by hassanms** [comment]
> @jamigibbs did [S3s CORS policies](https://docs.aws.amazon.com/AmazonS3/latest/dev/cors.html) help at all here? Having a similar issue to what you're experiencing and wondering if it's worth a shot
> 
> Edit: Success! 
> 
> I created a bucket with public access (it will act as a CDN), uploaded my image and making sure it was publicly accessible.
> 
> Then I added the following CORS configuration
> ![image](https://user-images.githubusercontent.com/9693812/91154445-e2d95200-e704-11ea-9d90-0dc5e39d1d32.png)
> 
> ```
> <CORSConfiguration>
>  <CORSRule>
>    <AllowedOrigin>*</AllowedOrigin>
>    <AllowedMethod>GET</AllowedMethod>
>  </CORSRule>
> </CORSConfiguration>
> ```
> 
> Worked like a charm.

JSON version:

[
    {
        "AllowedHeaders": [
            "*"
        ],
        "AllowedMethods": [
            "GET"
        ],
        "AllowedOrigins": [
            "*"
        ],
        "ExposeHeaders": []
    }
]

**Comment by Prajjwalgavander** [comment]
> I had the same CORS issue while accessing image stored in S3 bucket, I was able to access the image in normal `img` tag of browser but when the React-PDF tried to render it was throwing CORS error. Passing `"Cache-Control": "no-cache"` in Header worked for me
> 
> i.e while mentioning URL in Image tab of React-PDF instead of assigning URL directly to `src` attribute I made call as below
> `<Image src={{ uri: _your_image_url_goes_here_, method: "GET", headers: { "Cache-Control": "no-cache" }, body: "" }} />` and it worked.

Yes, this worked for my images stored in S3 bucket too. This is much more like bypassing CORS by using CORS-safelisted-response-header. Check this for reference - https://developer.mozilla.org/en-US/docs/Glossary/CORS-safelisted_response_header

**Comment by diegomura** [comment]
Closing now. Most of these issues were CORS related and were mitigated. If there's anything else regarding image loading, please submit another issue. This one has become a bit hard to follow

**Comment by ryancarville** [comment]
Still getting this. Any pointers on what is wrong?  Using standard usage and syntax of
`<Image src={profilePhoto.url} />`
also tried this but same results of CORS.  
` <Image
       src={{ uri: profilePhoto.url, method: 'GET', headers: { "Cache-Control": "no-cache" }, body: '' }} 
       style={commonStyles.image}
   />`
   
   Pulling fro a public S3 Bucket

**Comment by bjkuncoro** [comment]
this cors issue make me crazy, but finally i found solution that work like a charm for me, 
the solution is by using 3rd party CORS api, that make us not get image from real URL but get it from the proxy cors URL, 
ex:
`https://admin.sipakader.org/storage/berkas/61/6106/6106051/6106051005/photo_611291820289010.jpg`
and then i am using `https://cors.bridged.cc/` as a CORS proxy
so this the solution that work for me:
`https://cors.bridged.cc/https://admin.sipakader.org/storage/berkas/61/6106/6106051/6106051005/photo_611291820289010.jpg`

**Comment by ryancarville** [comment]
Thanks @bjkuncoro will look into it but would prefer if the library addressed this since it is such a sizable use case.

**Comment by Nutriadmin** [comment]
Please check #2340 for how I've solved this issue by making my server convert images to base64. This will work even when loading images from an external website that has a CORS policy you cannot control.

**Comment by dgreasi** [comment]
I had the same while using the `Image` tag in the pdf, at first from google drive link and afterwards from Cloud Storage bucket.

The html `img` tag was working out of the box but the `Image` component of the lib was always throwing the `CORS ERROR`.

My solution was the get the image with an api request and transform the image to `base64`. I did that with the following:
```
export const getCompanyLogoService = async (url: string): Promise<string | null> => {
  const resp = await axios.get<ArrayBuffer>(url, {
    responseType: 'arraybuffer',
  });

  if (resp?.data) {
    return Buffer.from(resp.data).toString('base64');
  }

  return null;
};
```

I also had to configure my bucket in GCP to be public for my testing and used the following cors config which i was able to set with the following command `gsutil cors set './relative-path-to-json-file' gs://{name-of-bucket}`

The cors.json file:
```
[
  {
    "origin": ["*"],
    "method": ["GET"],
    "responseHeader": ["Content-Type"],
    "maxAgeSeconds": 3600
  }
]
```

All the above are part of a standalone web app in React.

**Comment by muikku** [comment]
We also have this issue, and changing aws settings did not help.
We solved it by using proxy. In our case pictures are small, maybe larger ones could be problematic for client.

Server example

```
.use('/proxy', (req, res) => {
 const url = req.url?.slice(1);
    const response = await axios({
      method: 'GET',
      url, 
      responseType: 'arraybuffer',
    });

    const buffer = Buffer.from(response.data, 'binary');
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.end(buffer, 'binary');
})

```
    
Client example
    

```
const [imageSrc, setImageSrc] = React.useState(undefined);

 React.useEffect(() => {
  fetch(imageUrl), {
    method: 'GET',
  })
    .then((response) => response.blob())
    .then((blob) => {
      const source = URL.createObjectURL(blob);
      setImageSrc(source);
    })
    .catch((er) => console.error('Error fetching image:', er));
}
}, []);

   <Image src={imageSrc} />

```

**Comment by shubham-jangid** [comment]
I've converted the image to Base64 and tried loading it into the react-pdf <Image> component like this:

const src = 'data:image/png;base64,UklGRogX...'

<Image 
  style={styles.image}
  src={src} 
/>`

But unfortunately, it's still not working. Could anyone help me troubleshoot this?

**Comment by shubham-jangid** [comment]
I've converted the image to Base64 and tried loading it into the `react-pdf` `<Image>` component like this:

```javascript
const src = 'data:image/png;base64,UklGRogX...'

<Image 
  style={styles.image}
  src={src} 
/>
```

But unfortunately, it's still not working. Could anyone help me troubleshoot this?
